tests/test_updater_delegation_graphs.py

```diff
@@ -0,0 +1,246 @@
+#!/usr/bin/env python
+
+# Copyright 2021, New York University and the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Test updating delegated targets roles with various
+delegation hierarchies"""
+
+import os
+import sys
+import tempfile
+import unittest
+from dataclasses import astuple, dataclass, field
+from typing import Iterable, List, Optional
+from unittest.mock import call, patch
+
+from tests import utils
+from tests.repository_simulator import RepositorySimulator
+from tuf.api.metadata import (
+    SPECIFICATION_VERSION,
+    TOP_LEVEL_ROLE_NAMES,
+    DelegatedRole,
+    Targets,
+)
+from tuf.ngclient import Updater
+
+
+@dataclass
+class TestDelegation:
+    delegator: str
+    rolename: str
+    keyids: List[str] = field(default_factory=list)
+    threshold: int = 1
+    terminating: bool = False
+    paths: List[str] = field(default_factory=lambda: ["*"])
+    path_hash_prefixes: Optional[List[str]] = None
+
+
+@dataclass
+class DelegationsTestCase:
+    """Describes a delegations graph as a list of delegations
+    and the expected order of traversal as 'visited_order'."""
+
+    delegations: List[TestDelegation]
+    visited_order: List[str]
+
+
+class TestDelegationsGraphs(unittest.TestCase):
+    """Test creating delegations graphs with different complexity
+    and successfully updating the delegated roles metadata"""
+
+    def setUp(self) -> None:
+        self.temp_dir = tempfile.TemporaryDirectory()
+        self.metadata_dir = os.path.join(self.temp_dir.name, "metadata")
+        self.targets_dir = os.path.join(self.temp_dir.name, "targets")
+        os.mkdir(self.metadata_dir)
+        os.mkdir(self.targets_dir)
+
+    def tearDown(self) -> None:
+        self.temp_dir.cleanup()
+
+    def _init_updater(self, sim: RepositorySimulator) -> Updater:
+        """Create a new Updater instance"""
+        return Updater(
+            self.metadata_dir,
+            "https://example.com/metadata/",
+            self.targets_dir,
+            "https://example.com/targets/",
+            sim,
+        )
+
+    def _init_repo(
+        self, delegations: List[TestDelegation]
+    ) -> RepositorySimulator:
+        """Create a new RepositorySimulator instance with 'delegations'"""
+        sim = RepositorySimulator()
+        spec_version = ".".join(SPECIFICATION_VERSION)
+
+        for d in delegations:
+            if d.rolename in sim.md_delegates:
+                targets = sim.md_delegates[d.rolename].signed
+            else:
+                targets = Targets(1, spec_version, sim.safe_expiry, {}, None)
+
+            # unpack 'd' but skip "delegator"
+            role = DelegatedRole(*astuple(d)[1:])
```

Comment:
oh wow. mypy is going to have an aneurysm I bet :)

Reply:
Maybe it was paralysed by the shock because it stayed silent when I tried 😁